### PR TITLE
Clean up old migrations

### PR DESCRIPTION
**Pull request by esoergel**
Followup from https://github.com/dimagi/commcare-hq/pull/24725

For this PR, I looked at usages of `RunPython` and took one of three paths:

* Add `@skip_on_fresh_install`
  * Migration clearly migrates data without changing schemas, and is relatively new (written with the last year)
* ~Convert to a noop~ Deleted
  * Same as the above, but for old migrations
* Leave alone
  * Touches schema or db partitioning
  * Bootstraps data (I pretty much ignored SMS and accounting migrations)
  * Questionable or unclear action 

I'd appreciate a thorough review here - I don't have direct context with most of these migrations, they looks safe to remove or skip, but I'd like a second opinion to make sure I'm not overlooking something.

Read commit-by-commit.

### CODE REVIEW

File: corehq/util/django_migrations.py [hunk 1]
```diff
@@ -99,3 +100,10 @@ def _inner(*args, **kwargs):
             return
         return migration_fn(*args, **kwargs)
     return _inner
+
+
+def noop_migration():
+    """
+    A migration that does nothing. Used to replace old migrations that are no longer required e.g moved.
+    """
+    return RunPython(RunPython.noop, RunPython.noop)
```
Comment (esoergel) [inline corehq/util/django_migrations.py]: Should this be [`elidable=True`](https://docs.djangoproject.com/en/2.2/ref/migration-operations/#runpython)?

File: corehq/apps/data_interfaces/migrations/0018_check_for_rule_migration.py [hunk 2]
```diff
@@ -3,24 +3,7 @@
...
-    if AutomaticUpdateRule.objects.filter(migrated=False).count() > 0:
-        raise RuntimeError("""
-            If you are receiving this error in a Dimagi production environment,
-            it is unexpected.
-
-            Otherwise, you have unmigrated AutomaticUpdateRules. You can migrate them
-            by checking out tag `rule-migration` and running:
-
-                python manage.py migrate_rules
-        """)
```
Comment (esoergel) [inline corehq/apps/data_interfaces/migrations/0018_check_for_rule_migration.py]: This one was on the fence with my self-imposed "one year" rule (it's from September 2018).  Given that it requires you to check out an old branch and run it (something which gets progressively harder over time), I figured I'd delete it.

Aside - this looks like all `AutomaticUpdateRule`s have been deleted, could probably delete that model and associated code now, but I'm not gonna go down that route.

File: corehq/apps/case_search/migrations/0008_auto_20180119_1716.py [hunk 3]
```diff
@@ -1,14 +1,14 @@
...
+from __future__ import absolute_import, unicode_literals
 
 from django.core.management import call_command
 from django.db import migrations
 
+from corehq.util.django_migrations import skip_on_fresh_install
 
 
-
+@skip_on_fresh_install
```
Comment (snopoke) [inline corehq/apps/case_search/migrations/0008_auto_20180119_1716.py]: I think this can be made no-op

File: corehq/sql_accessors/migrations/0019_get_case_ids_in_domain_by_owners.py [hunk 4]
```diff
@@ -4,7 +4,8 @@
 from __future__ import absolute_import
 from django.db import migrations
 
-from corehq.sql_db.operations import noop_migration, RawSQLMigration
+from corehq.sql_db.operations import RawSQLMigration
+from corehq.util.django_migrations import noop_migration
 
 migrator = RawSQLMigration(('corehq', 'sql_accessors', 'sql_templates'), {})
```
Comment (snopoke) [inline corehq/sql_accessors/migrations/0019_get_case_ids_in_domain_by_owners.py]: this can be removed as well as the import

File: corehq/sql_accessors/migrations/0023_update_get_case_form_ids.py [hunk 5]
```diff
@@ -5,7 +5,8 @@
 from django.db import migrations
 
 from corehq.form_processor.models import CaseTransaction
-from corehq.sql_db.operations import RawSQLMigration, noop_migration
+from corehq.sql_db.operations import RawSQLMigration
+from corehq.util.django_migrations import noop_migration
 
 
 migrator = RawSQLMigration(('corehq', 'sql_accessors', 'sql_templates'), {
```
Comment (snopoke) [inline corehq/sql_accessors/migrations/0023_update_get_case_form_ids.py]: can remove this too

File: corehq/sql_accessors/migrations/0038_attachment_bucket.py [hunk 6]
```diff
@@ -4,7 +4,8 @@
 from __future__ import absolute_import
 from django.db import migrations
 
-from corehq.sql_db.operations import RawSQLMigration, noop_migration
+from corehq.sql_db.operations import RawSQLMigration
+from corehq.util.django_migrations import noop_migration
 
 migrator = RawSQLMigration(('corehq', 'sql_accessors', 'sql_templates'), {})
```
Comment (snopoke) [inline corehq/sql_accessors/migrations/0038_attachment_bucket.py]: another for removal

File: corehq/util/django_migrations.py (hunk 1 above)
Comment (snopoke) [reply corehq/util/django_migrations.py]: yes

File: custom/icds_reports/migrations/0021_remove_categories.py [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 from __future__ import absolute_import
 from django.db import migrations, models
-from corehq.sql_db.operations import noop_migration
+from corehq.util.django_migrations import noop_migration
 from corehq.sql_db.operations import RawSQLMigration
 
 migrator = RawSQLMigration(('custom', 'icds_reports', 'migrations', 'sql_templates'))
```
Comment (snopoke) [inline custom/icds_reports/migrations/0021_remove_categories.py]: :skull:

File: custom/icds_reports/migrations/0023_fix_view.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 
 from __future__ import absolute_import
 from django.db import migrations, models
-from corehq.sql_db.operations import noop_migration
+from corehq.util.django_migrations import noop_migration
 from corehq.sql_db.operations import RawSQLMigration
 
 migrator = RawSQLMigration(('custom', 'icds_reports', 'migrations', 'sql_templates'))
```
Comment (snopoke) [inline custom/icds_reports/migrations/0023_fix_view.py]: :skull:

File: custom/icds_reports/migrations/0025_add_fully_immunized_to_view.py [hunk 9]
```diff
@@ -3,7 +3,7 @@
 
 from __future__ import absolute_import
 from django.db import migrations, models
-from corehq.sql_db.operations import noop_migration
+from corehq.util.django_migrations import noop_migration
 from corehq.sql_db.operations import RawSQLMigration
 
 migrator = RawSQLMigration(('custom', 'icds_reports', 'migrations', 'sql_templates'))
```
Comment (snopoke) [inline custom/icds_reports/migrations/0025_add_fully_immunized_to_view.py]: :skull:

File: custom/icds_reports/migrations/0042_child_health_monthly_view.py [hunk 10]
```diff
@@ -4,7 +4,8 @@
 from __future__ import unicode_literals
 
 from django.db import migrations
-from corehq.sql_db.operations import RawSQLMigration, noop_migration
+from corehq.sql_db.operations import RawSQLMigration
+from corehq.util.django_migrations import noop_migration
 
 migrator = RawSQLMigration(('custom', 'icds_reports', 'migrations', 'sql_templates'))
```
Comment (snopoke) [inline custom/icds_reports/migrations/0042_child_health_monthly_view.py]: :skull:

File: corehq/sql_db/operations.py [hunk 11]
```diff
@@ -1,30 +1,25 @@
...
 
 class IndexRenameOperationException(Exception):
     pass
 
 
-def noop_migration():
-    """
-    A migration that does nothing. Used to replace old migrations that are no longer required e.g moved.
-    """
-    return RunPython(RunPython.noop, RunPython.noop)
```
Comment (millerdev) [inline corehq/sql_db/operations.py]: Why was this moved to `corehq.util.django_migrations`? It created a lot of noise in this PR.

File: corehq/motech/repeaters/migrations/0001_adjust_auth_field_format.py [hunk 12]
```diff
@@ -1,31 +1,14 @@
...
+from corehq.util.django_migrations import noop_migration
 
 
 class Migration(migrations.Migration):
 
     dependencies = []
 
     operations = [
-        migrations.RunPython(migrate_auth_field),
+        noop_migration(),
```
Comment (emord) [inline corehq/motech/repeaters/migrations/0001_adjust_auth_field_format.py]: I think that these don't actually need to be `noop_migration`. I believe that they can be 

```operations = []```

File: corehq/sql_db/operations.py (hunk 11 above)
Comment (esoergel) [reply corehq/sql_db/operations.py]: The concept of a noop migration isn't specific to SQL db operations - most of the places I added it in this PR weren't related to SQL models at all.  It seemed like the new location made more sense, and I wanted it to be right next to `skip_on_fresh_install`, since they're closely related (I imagine most usages of `skip_on_fresh_install` will eventually be converted to `noop_migration()`s.

I did briely consider importing it to `operations.py` so I wouldn't have to update the import statements, but figured that'd just be laziness.  I committed the move separately to hopefully make the noise ignorable, at least.

File: corehq/motech/repeaters/migrations/0001_adjust_auth_field_format.py (hunk 12 above)
Comment (esoergel) [reply corehq/motech/repeaters/migrations/0001_adjust_auth_field_format.py]: I thought Django required an operation, was going off of this:
https://github.com/dimagi/commcare-hq/blob/f3f04690be1581c9e11806df2a3cd56c166ca6d4/corehq/form_processor/README.md#L98-L100

Are you saying there's never any reason to use `noop_migration()`?

Comment (esoergel) [reply corehq/motech/repeaters/migrations/0001_adjust_auth_field_format.py]: offlined with @emord.  Looks like that is indeed the case.  Will update this PR accordingly.

File: corehq/util/django_migrations.py (hunk 1 above)
Comment (esoergel) [reply corehq/util/django_migrations.py]: At @emord 's suggestion, I removed nearly all `noop_migration`s, save a few which used this dynamically.  Those ones I think _don't_ want `elidable=True`, since whether the migration is actually a noop depends on the settings.  I changed the docstring to reflect the intended usage.

**Comment by esoergel** [comment]
Phew, well that was tedious.  I'd absolutely recommend reviewing individual commits to make sense of this PR.  There are two commits with very big (~100 file) bulk changes which should all be pretty rote.  I did the bulk of it with text editor macros and isort.

**Review by emord — APPROVED** [review]
> recommend reviewing individual commits to make sense of this PR

I think I'd recommend against this. It was way easier for me to go to the files and check "viewed" to keep track of the files and my progress (I believe it is saved between page reloads as well). not sure if that helps if you update this PR.

> I did the bulk of it with text editor macros and isort.

FWIW I've been following a process where I've been including what I actually did in the commit message or body. So when I isort a file(s), that's all I do in that commit and the message is "isort filepath". Or when I do some nasty bash thing the body looks like: https://github.com/dimagi/commcare-hq/commit/39891b784d6a4f1e2b581d4f6458b7ecc2a2e41b

File: corehq/form_processor/README.md [hunk 13]
```diff
@@ -95,12 +94,10 @@ example](https://github.com/dimagi/commcare-hq/pull/19195):
...
-    happening in the migration.
+3. Remove all references to the old file in existing migrations
 4. Repeat the above for the pl_proxy function defined in `sql_proxy_accessors`
    if applicable.
 
-These `migrations.RunSQL` operations can later be replaced with `noop_migration`s, but
-there's no urgency in that, so we can wait until someone goes through and
-squashes all these migrations.
+These `migrations.RunSQL` operations can later be deleted, but there's no
+urgency in that, so we can wait until someone goes through and squashes all
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/form_processor/README.md]: suggestion vehemently rejected

**Comment by esoergel** [comment]
https://github.com/dimagi/commcare-hq/pull/24735#pullrequestreview-259721243
Both excellent suggestions, thanks @emord !

**Review by millerdev — APPROVED** [review]

**Comment by emord** [comment]
potential addition to flaky tests
```

======================================================================
FAIL: corehq.apps.api.tests.user_resources:TestCommCareUserResource.test_get_single
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/api/tests/user_resources.py", line 50, in test_get_single
    self.assertEqual(response.status_code, 200)
AssertionError: 403 != 200

```

**Comment by emord** [comment]
and another: 

```
======================================================================
ERROR: corehq.apps.api.odata.tests.test_metadata:TestDeprecatedFormMetadataDocumentWithTwoFactorUsingApiKey.test_wrong_password
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/api/odata/tests/test_metadata.py", line 153, in test_wrong_password
    response = self._execute_query(wrong_credentials)
  File "/mnt/commcare-hq-ro/corehq/apps/api/odata/tests/utils.py", line 64, in _execute_query
    return self.client.get(self.view_url, HTTP_AUTHORIZATION='Basic ' + credentials)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 536, in get
    **extra)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 340, in get
    return self.generic('GET', path, secure=secure, **r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 416, in generic
    return self.request(**r)
  File "/vendor/lib/python2.7/site-packages/django/test/client.py", line 501, in request
    six.reraise(*exc_info)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/vendor/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/vendor/lib/python2.7/site-packages/django/utils/decorators.py", line 67, in _wrapper
    return bound_func(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 199, in safe_fn
    return _inner(request, domain, *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/domain/decorators.py", line 411, in _inner
    return fn(request, *args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/domain/auth.py", line 157, in wrapper
    user = authenticate(username=username, password=password, request=request)
  File "/vendor/lib/python2.7/site-packages/django/contrib/auth/__init__.py", line 81, in authenticate
    user_login_failed.send(sender=__name__, credentials=_clean_credentials(credentials), request=request)
  File "/vendor/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 193, in send
    for receiver in self._live_receivers(sender)
  File "/mnt/commcare-hq-ro/corehq/apps/hqwebapp/signals.py", line 34, in add_failed_attempt
    user.save()
  File "/mnt/commcare-hq-ro/corehq/apps/users/models.py", line 1536, in save
    super(CouchUser, self).save(**params)
  File "/vendor/lib/python2.7/site-packages/couchdbkit/schema/base.py", line 134, in save
    db.save_doc(doc, **params)
  File "/vendor/lib/python2.7/site-packages/couchdbkit/client.py", line 533, in save_doc
    raise ResourceConflict
ResourceConflict: 
```

**Comment by esoergel** [comment]
@dimagi/team-commcare-hq This PR attempts to standardize a couple aspects of data migrations in Django.

### Short description
Use `@skip_on_fresh_install` to skip migrations which only need to run on existing, populated databases, and not in tests or new environments.  `noop_migration()` is usually not necessary - you can just delete old `RunPython` instantiations.

### Long description
We used to frequently use management commands to update existing data, which gives us control of when a potentially lengthy migration is run.  However, with an increasing number of environments outside of our control, this isn't a tenable solution alone, so we have lately been writing these as django migrations instead, or as management commands that are are also called by the migration.

You may use ```@skip_on_fresh_install``` up-front on your migration function to skip it for tests and new environments, so they don't need to hit the db to see that there are no docs needing a migration.

These migrations need to be kept around until all environments which need them have been migrated.  I don't have a good sense of how long that actually is - if you haven't ran `./manage.py migrate` in a year, can we assume you're not that concerned about keeping an up-to-date, functional environment?

At any rate, once it's been determined that it's safe to remove the migration code, you can just delete the `RunPython` instantiation and the code it references.  No need to replace it with `noop_migration()`.   This is a perfectly valid migration:
```python
class Migration(migrations.Migration):
    dependencies = [...]
    operations = []
```
